wait_us optimization

kjbracey opened the pull request:
### Description

As the timer code became more generic, coping with initialization on demand, and variable width and speed us_ticker_api implementations, `wait_us` has gradually gotten slower and slower.

Some platforms have reportedly seen overhead of `wait_us()` increase from 10µs to 30µs. These changes should fully reverse that drop, and even make it better than ever.

Add fast paths for platforms that provide compile-time information about `us_ticker`. Speed and code size is improved further if:

* Timer has >= 2^32 microsecond range, or better still is 32-bit 1MHz.
* Platform implements `us_ticker_read()` as a macro
* Timer is initialised at boot, rather than first use

The latter initialisation option is the default for STM, as this has always been the case.

PR adds support for the optimization for Freescale MCUXpresso family and all STM devices.

Testing on K64F, a tight `{ led = !led; wait_us(10); }` loop does achieve 10µs transitions with this change. Without, it was 16µs.

Information provided by targets is only currently used to optimize `wait_us`, not any other use of us_ticker via `Ticker` et al, but the information could be used to optimize those in future.

### Pull request type

    [ ] Fix
    [X] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@bulislaw 

### Release Notes

* `wait_us` has been optimized for certain platforms - see `us_ticker_api.h` for the details of enabling optimizations on other targets. It can be optimized further if the option `target.init-us-ticker-at-boot` is enabled.

kjbracey committed:
wait_us optimization

As the timer code became more generic, coping with initialization on
demand, and variable width and speed us_ticker_api implementations,
wait_us has gradually gotten slower and slower.

Some platforms have reportedly seen overhead of wait_us() increase from
10µs to 30µs. These changes should fully reverse that drop, and even
make it better than ever.

Add fast paths for platforms that provide compile-time information about
us_ticker. Speed and code size is improved further if:

* Timer has >= 2^32 microsecond range, or better still is 32-bit 1MHz.
* Platform implements us_ticker_read() as a macro
* Timer is initialised at boot, rather than first use

The latter initialisation option is the default for STM, as this has
always been the case.

kjbracey committed:
STM: optimize us_ticker

Provide new optimizations for us_ticker and wait_us.

kjbracey committed:
Freescale MCUXpresso: optimize us_ticker

Provide new optimizations for us_ticker and wait_us.

bulislaw reviewed (approved):
Looks good, is there a point of doing something similar for LP ticker?

pan- reviewed (commented):
The change looks good however it is confusing to have macros and functions with the same name. Wouldn't it be better to mark the default implementation of `us_ticker_read` as weak and override it as a `static inline` function in header files of platform supporting this optimisation ?

jeromecoutant reviewed (approved):
Non regression executed with ['NUCLEO_F091RC', 'NUCLEO_F103RB', 'NUCLEO_F207ZG', 'NUCLEO_F303ZE', 'NUCLEO_F446RE', 'NUCLEO_F767ZI', 'NUCLEO_H743ZI', 'NUCLEO_L073RZ', 'NUCLEO_L152RE', 'NUCLEO_L476RG', 'NUCLEO_WB55RG']

kjbracey committed:
wait_ns micro-optimisation

Use `+ 1` to set the "Thumb" indicator on the `delay_loop` routine
address, rather than `| 1`. That makes it something that can be done
by the linker, rather than needing to be done at run-time.

Saves one instruction and one cycle.

bulislaw reviewed (approved):
Could you make sure that porting guide details (step by step) how to implement this optimised version of us ticker.

kjbracey committed:
Remote "static" from MBED_FORCEINLINE

Static keyword causes problems when trying to use force-inlined
functions from normal inlined functions. This is not legal:

    static inline void forced() { }

    inline void normal() { forced(); }

You cannot reference internal-linkage things from external-linkage
inline functions.

Removal of the static implies that in C there would need to be a
non-inline definition in case anyone calls it non-inlined, but if the
force attribute is doing its job, that should not happen.

Only significant in-tree user of the MBED_FORCEINLINE macro is
the atomic operations - making this change permits atomic operations
from non-static inline functions.
